File: src/algorithms/arrays/minimum-subarray-sum/sources/minimum-subarray-sum.ts
// Minimum Subarray Sum — O(n) inverted Kadane's algorithm tracking minimum instead of maximum
function minimumSubarraySum(inputArray: number[]): MinimumSubarraySumResult {
  if (inputArray.length === 0) {
    // @step:initialize
    return { minSum: 0, startIndex: 0, endIndex: 0 }; // @step:initialize
  }

  let minEndingHere = inputArray[0]!; // @step:initialize
  let minSoFar = inputArray[0]!; // @step:initialize
  let currentStartIndex = 0;
  let bestStartIndex = 0;
  let bestEndIndex = 0;

  // Extend the current subarray or restart from the current element
  for (let elementIndex = 1; elementIndex < inputArray.length; elementIndex++) {
    if (inputArray[elementIndex]! < minEndingHere + inputArray[elementIndex]!) {
      // @step:compare
      minEndingHere = inputArray[elementIndex]!; // @step:compare
      currentStartIndex = elementIndex; // @step:compare
    } else {
      minEndingHere += inputArray[elementIndex]!; // @step:compare
    }

    if (minEndingHere < minSoFar) {
      // @step:compare
      minSoFar = minEndingHere; // @step:compare
      bestStartIndex = currentStartIndex; // @step:compare
      bestEndIndex = elementIndex; // @step:compare
    }
  }

  return { minSum: minSoFar, startIndex: bestStartIndex, endIndex: bestEndIndex }; // @step:complete
}

File: src/algorithms/arrays/sliding-window/minimum-subarray-sum/sources/minimum-subarray-sum.go
// Minimum Subarray Sum — O(n) inverted Kadane's algorithm tracking minimum instead of maximum
package minimumsubarraysum

func minimumSubarraySum(inputArray []int) (minSum int, startIndex int, endIndex int) {
	if len(inputArray) == 0 {
		// @step:initialize
		return 0, 0, 0 // @step:initialize
	}

	minEndingHere := inputArray[0] // @step:initialize
	minSoFar := inputArray[0]      // @step:initialize
	currentStartIndex := 0
	bestStartIndex := 0
	bestEndIndex := 0

	// Extend the current subarray or restart from the current element
	for elementIndex := 1; elementIndex < len(inputArray); elementIndex++ {
		if inputArray[elementIndex] < minEndingHere+inputArray[elementIndex] { // @step:compare
			minEndingHere = inputArray[elementIndex]  // @step:compare
			currentStartIndex = elementIndex          // @step:compare
		} else {
			minEndingHere += inputArray[elementIndex] // @step:compare
		}

		if minEndingHere < minSoFar { // @step:compare
			minSoFar = minEndingHere          // @step:compare
			bestStartIndex = currentStartIndex // @step:compare
			bestEndIndex = elementIndex        // @step:compare
		}
	}

	return minSoFar, bestStartIndex, bestEndIndex // @step:complete
}
